sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetColumnVector.java

```diff
@@ -0,0 +1,379 @@
+    // nothing to do if the column itself is missing
+    if (vector.isAllNull()) return;
+
+    DataType type = column.sparkType();
+    if (type instanceof ArrayType || type instanceof MapType) {
+      for (ParquetColumnVector child : children) {
+        child.assemble();
+      }
+      calculateCollectionOffsets();
+    } else if (type instanceof StructType) {
+      for (ParquetColumnVector child : children) {
+        child.assemble();
+      }
+      calculateStructOffsets();
+    }
+  }
+
+  /**
+   * Resets this Parquet column vector, which includes resetting all the writable column vectors
+   * (used to store values, definition levels, and repetition levels) for this and all its children.
+   */
+  void reset() {
+    // nothing to do if the column itself is missing
+    if (vector.isAllNull()) return;
+
+    vector.reset();
+    repetitionLevels.reset();
+    definitionLevels.reset();
+    for (ParquetColumnVector child : children) {
+      child.reset();
+    }
+  }
+
+  /**
+   * Returns the {@link ParquetColumn} of this column vector.
+   */
+  ParquetColumn getColumn() {
+    return this.column;
+  }
+
+  /**
+   * Returns the writable column vector used to store values.
+   */
+  WritableColumnVector getValueVector() {
+    return this.vector;
+  }
+
+  /**
+   * Returns the writable column vector used to store repetition levels.
+   */
+  WritableColumnVector getRepetitionLevelVector() {
+    return this.repetitionLevels;
+  }
+
+  /**
+   * Returns the writable column vector used to store definition levels.
+   */
+  WritableColumnVector getDefinitionLevelVector() {
+    return this.definitionLevels;
+  }
+
+  /**
+   * Returns the column reader for reading a Parquet column.
+   */
+  VectorizedColumnReader getColumnReader() {
+    return this.columnReader;
+  }
+
+  /**
+   * Sets the column vector to 'reader'. Note this can only be called on a primitive Parquet
+   * column.
+   */
+  void setColumnReader(VectorizedColumnReader reader) {
+    if (!isPrimitive) {
+      throw new IllegalStateException("Can't set reader for non-primitive column");
+    }
+    this.columnReader = reader;
+  }
+
+  private void calculateCollectionOffsets() {
+    int maxDefinitionLevel = column.definitionLevel();
+    int maxElementRepetitionLevel = column.repetitionLevel();
+
+    ParquetColumn child = column.children().apply(0);
+    int maxElementDefinitionLevel = child.definitionLevel();
+    if (!child.required()) maxElementDefinitionLevel -= 1;
+
+    // There are 4 cases when calculating definition levels:
+    //   1. definitionLevel == maxDefinitionLevel
+    //     ==> value is defined and not null
+    //   2. definitionLevel == maxDefinitionLevel - 1
+    //     ==> value is null
+    //   3. definitionLevel < maxDefinitionLevel - 1
+    //     ==> value doesn't exist since one of its optional parents is null
+    //   4. definitionLevel > maxDefinitionLevel
+    //     ==> value is a nested element within an array or map
```

Comment:
Any example for case 4? You mean the `maxDefinitionLevel` is the definition level of the array? Then isn't the `maxDefinitionLevel`  should be `array's definitionLevel + 1`

Reply:
For example, an optional array with optional element is represented by:
```
                                        RL   DL
message schema {                        0    0
  optional group _1 (LIST) {            0    1
    repeated group list {               1    2
      optional binary element (UTF8);   1    3
    }
  }
}
```

here `RL` and `DL` are the max repetition and definition level for the schema element at the corresponding positions. The `maxDefinitionLevel` here is the max definition level of the array itself, which is 1. Any array element should have definition level > 1.

Reply:
Oh, I see. `maxDefinitionLevel` is confused me that as I thought it is max def level for elements. Because this equation `definitionLevel > maxDefinitionLevel` looks not reasonable. Actually you indicate two def levels at different element (array and nested element).

Reply:
Yea `maxDefinitionLevel` and `maxRepetitionLevel` is for the collection itself here. For collection elements, the naming would be `maxElementDefinitionLevel` and `maxElementRepetitionLevel`.